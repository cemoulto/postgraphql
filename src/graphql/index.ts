import createGraphqlSchema from './schema/createGraphqlSchema'

export {
  createGraphqlSchema,
}
