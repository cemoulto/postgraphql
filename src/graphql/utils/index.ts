import buildObject from './buildObject'
import formatName from './formatName'
import idSerde from './idSerde'

export { buildObject, formatName, idSerde }

export * from './memoize'
