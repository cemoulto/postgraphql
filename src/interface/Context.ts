type CleanupFn = () => void | Promise<void>

/**
 * `Context` is an object which holds information about any single logical data
 * access.
 */
// TODO: tests
class Context {
  /**
   * Dependencies which can be injected into functions. Often different drivers
   * will implement their own getters/setters to maintain type information.
   *
   * Uses symbols to prevent key collisions.
   *
   * @private
   */
  private _dependencies = new Map<Symbol, mixed>()

  /**
   * Adds a dependency using a symbol as the key. If a dependency with that
   * symbol already exists, an error will be thrown.
   *
   * @see `Context#getDependency`
   */
  public addDependency <T>(symbol: Symbol, dependency: T): this {
    if (this._dependencies.has(symbol))
      throw new Error('A dependency for this symbol already exists.')

    this._dependencies.set(symbol, dependency)

    return this
  }

  /**
   * Gets a dependency by its symbol and returns it. If no such dependency
   * for the symbol exists, the method returns `undefined`.
   *
   * @see `Context#addDependency`
   */
  public getDependency <T>(symbol: Symbol): T | undefined {
    return this._dependencies.get(symbol) as any
  }

  /**
   * An array of the cleanup functions that have been added to our context.
   *
   * @private
   */
  private _cleanupFns: Array<CleanupFn> = []

  /**
   * Adds a cleanup function which will be run when `Context#cleanup` is
   * called.
   *
   * @see `Context#cleanup`
   */
  public addCleanupFunction (cleanupFn: CleanupFn): this {
    this._cleanupFns.push(cleanupFn)
    return this
  }

  /**
   * A method which *must* be called whenever a context has finished its task.
   *
   * @see `Context#addCleanupFunction`
   */
  public async cleanup () {
    await Promise.all(this._cleanupFns.map(cleanupFn => cleanupFn()))
  }
}

export default Context
