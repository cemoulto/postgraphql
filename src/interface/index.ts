import Inventory from './Inventory'
import Collection from './collection/Collection'
import CollectionKey from './collection/CollectionKey'
import Paginator from './collection/Paginator'
import Condition from './collection/Condition'
import Relation from './collection/Relation'
import Type from './type/Type'
import NullableType from './type/NullableType'
import ListType from './type/ListType'
import NamedType, { isNamedType } from './type/NamedType'
import AliasType from './type/AliasType'
import EnumType from './type/EnumType'
import booleanType from './type/primitive/booleanType'
import integerType from './type/primitive/integerType'
import floatType from './type/primitive/floatType'
import stringType from './type/primitive/stringType'
import ObjectType from './type/ObjectType'

export {
  Inventory,
  Collection,
  CollectionKey,
  Paginator,
  Condition,
  Relation,
  Type,
  NullableType,
  ListType,
  NamedType,
  isNamedType,
  AliasType,
  EnumType,
  booleanType,
  integerType,
  floatType,
  stringType,
  ObjectType,
}
