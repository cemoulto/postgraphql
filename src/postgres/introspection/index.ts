import introspectDatabase from './introspectDatabase'
import PGCatalog from './PGCatalog'
import PGObject from './object/PGObject'
import PGNamespaceObject from './object/PGNamespaceObject'
import PGClassObject from './object/PGClassObject'
import PGAttributeObject from './object/PGAttributeObject'
import PGTypeObject from './object/PGTypeObject'

export {
  introspectDatabase,
  PGCatalog,
  PGObject,
  PGNamespaceObject,
  PGClassObject,
  PGAttributeObject,
  PGTypeObject,
}
