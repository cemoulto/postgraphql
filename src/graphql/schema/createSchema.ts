import { GraphQLSchema, GraphQLObjectType, GraphQLFieldConfig } from 'graphql'
import { Inventory } from '../../interface'
import buildObject from '../utils/buildObject'
import createNodeFieldEntry from './node/createNodeFieldEntry'
import getCollectionType from './collection/getCollectionType'
import createCollectionQueryFieldEntries from './collection/createCollectionQueryFieldEntries'
import BuildToken from './BuildToken'

type Options = {
  nodeIdFieldName?: string,
}

// TODO: doc
export default function createSchema (inventory: Inventory, options: Options = {}): GraphQLSchema {
  // We take our user-friendly arguments to `createSchema` and convert them
  // into a build token. One nice side effect of always creating our own
  // build token object is that we have the guarantee that every build token
  // will always maintain its own memoization map.
  const buildToken: BuildToken = {
    inventory,
    options: {
      // The default node id field name is `__id` as it is the emerging
      // standard.
      nodeIdFieldName: options.nodeIdFieldName || '__id',
    },
  }

  return new GraphQLSchema({
    query: createQueryType(buildToken),
    types: [
      // Make sure to always include the types for our collections, even if
      // they have no other output.
      ...inventory.getCollections().map(collection => getCollectionType(buildToken, collection)),
    ],
  })
}

// TODO: doc
function createQueryType <T>(buildToken: BuildToken): GraphQLObjectType<T> {
  const { inventory } = buildToken
  return new GraphQLObjectType({
    name: 'Query',
    // TODO: description
    fields: buildObject<GraphQLFieldConfig<T, mixed>>(
      [
        createNodeFieldEntry(buildToken),
      ],
      inventory
        .getCollections()
        .map(collection => createCollectionQueryFieldEntries(buildToken, collection))
        .reduce((a, b) => a.concat(b), []),
    ),
  })
}
