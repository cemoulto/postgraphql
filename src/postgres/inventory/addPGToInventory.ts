import { InventoryÂ } from '../../interface'
import PGCatalog from '../introspection/PGCatalog'
import PGCollection from './collection/PGCollection'

/**
 * Adds Postgres based objects created by introspection to an inventory.
 */
export default function addPGToInventory (inventory: Inventory, pgCatalog: PGCatalog) {
  // Add all of our collections.
  for (const pgClass of pgCatalog.getClasses())
    inventory.addCollection(new PGCollection(pgCatalog, pgClass))
}
