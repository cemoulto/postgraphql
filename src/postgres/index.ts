import addPGToInventory from './addPGToInventory'

export { addPGToInventory }
