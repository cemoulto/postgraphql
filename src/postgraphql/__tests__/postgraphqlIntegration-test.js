import { printSchema } from 'graphql'
import { Inventory } from '../../interface'
import getTestPGClient from '../../postgres/__tests__/fixtures/getTestPGClient'
import { introspectDatabase } from '../../postgres/introspection'
import addPGToInventory from '../../postgres/inventory/addPGToInventory'
import createSchema from '../../graphql/schema/createSchema'

test('forum example', async () => {
  const client = await getTestPGClient()
  const catalog = await introspectDatabase(client, ['a', 'b', 'c'])
  const inventory = new Inventory()
  addPGToInventory(inventory, catalog)
  const schema = createSchema(inventory)

  expect(printSchema(schema)).toMatchSnapshot()
})
