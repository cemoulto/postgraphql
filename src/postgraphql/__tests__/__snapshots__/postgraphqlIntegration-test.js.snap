exports[`test query relation-tail-head.graphql 1`] = `
Object {
  "data": Object {
    "allCompoundKeys": Object {
      "nodes": Array [
        Object {
          "extra": null,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "john.smith@email.com",
            "name": "John Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personId1": 1,
          "personId2": 2
        },
        Object {
          "extra": false,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "john.smith@email.com",
            "name": "John Smith"
          },
          "personId1": 2,
          "personId2": 1
        },
        Object {
          "extra": null,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "budd.deey@email.com",
            "name": "Budd Deey"
          },
          "personId1": 2,
          "personId2": 3
        },
        Object {
          "extra": true,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "sara.smith@email.com",
            "name": "Sara Smith"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "johnny.tucker@email.com",
            "name": "Johnny Tucker"
          },
          "personId1": 2,
          "personId2": 5
        },
        Object {
          "extra": true,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "budd.deey@email.com",
            "name": "Budd Deey"
          },
          "personId1": 4,
          "personId2": 3
        },
        Object {
          "extra": false,
          "personByPersonId1": Object {
            "createdAt": null,
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez"
          },
          "personByPersonId2": Object {
            "createdAt": null,
            "email": "kathryn.ramirez@email.com",
            "name": "Kathryn Ramirez"
          },
          "personId1": 4,
          "personId2": 4
        }
      ]
    }
  }
}
`;

exports[`test schema 1`] = `
"schema {
  query: Query
}

scalar AnInt

scalar AnotherInt

enum Color {
  RED
  GREEN
  BLUE
}

type CompoundKey implements Node {
  __id: ID!
  personId2: Int!
  personId1: Int!
  extra: Boolean
  personByPersonId1: Person
  personByPersonId2: Person
}

type CompoundKeysConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [CompoundKeysEdge]
  nodes: [CompoundKey!]
}

type CompoundKeysEdge {
  cursor: Cursor!
  node: CompoundKey!
}

enum CompoundKeysOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  PERSON_ID_2_ASC
  PERSON_ID_2_DESC
  PERSON_ID_1_ASC
  PERSON_ID_1_DESC
  EXTRA_ASC
  EXTRA_DESC
}

type CompoundType {
  a: Int
  b: String
  c: Color
  d: Uuid
}

scalar Cursor

scalar Email

interface Node {
  __id: ID!
}

type NonUpdatableView {
  column: Int
}

type NonUpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [NonUpdatableViewsEdge]
  nodes: [NonUpdatableView!]
}

type NonUpdatableViewsEdge {
  cursor: Cursor!
  node: NonUpdatableView!
}

enum NonUpdatableViewsOrderBy {
  NATURAL
  COLUMN_ASC
  COLUMN_DESC
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: Cursor
  endCursor: Cursor
}

type PeopleConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PeopleEdge]
  nodes: [Person!]
}

type PeopleEdge {
  cursor: Cursor!
  node: Person!
}

enum PeopleOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  ABOUT_ASC
  ABOUT_DESC
  EMAIL_ASC
  EMAIL_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
}

type Person implements Node {
  __id: ID!
  id: Int!
  name: String!
  about: String
  email: Email
  createdAt: String
}

type Post implements Node {
  __id: ID!
  id: Int!
  headline: String!
  body: String
  authorId: Int
  personByAuthorId: Person
}

type PostsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PostsEdge]
  nodes: [Post!]
}

type PostsEdge {
  cursor: Cursor!
  node: Post!
}

enum PostsOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  HEADLINE_ASC
  HEADLINE_DESC
  BODY_ASC
  BODY_DESC
  AUTHOR_ID_ASC
  AUTHOR_ID_DESC
}

type Query {
  node(__id: ID!): Node
  allNonUpdatableViews(orderBy: NonUpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): NonUpdatableViewsConnection
  allPosts(orderBy: PostsOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PostsConnection
  post(__id: ID!): Post
  postById(id: Int!): Post
  allTypes(orderBy: TypesOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): TypesConnection
  allUpdatableViews(orderBy: UpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): UpdatableViewsConnection
  allCompoundKeys(orderBy: CompoundKeysOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): CompoundKeysConnection
  compoundKey(__id: ID!): CompoundKey
  compoundKeyByPersonId1AndPersonId2(personId1: Int!, personId2: Int!): CompoundKey
  allPeople(orderBy: PeopleOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PeopleConnection
  person(__id: ID!): Person
  personByEmail(email: Email): Person
  personById(id: Int!): Person
}

type Types {
  bigint: Int
  boolean: Boolean
  varchar: String
  enum: Color
  domain: AnInt
  domain2: AnotherInt
  compoundType: CompoundType
}

type TypesConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [TypesEdge]
  nodes: [Types!]
}

type TypesEdge {
  cursor: Cursor!
  node: Types!
}

enum TypesOrderBy {
  NATURAL
  BIGINT_ASC
  BIGINT_DESC
  BOOLEAN_ASC
  BOOLEAN_DESC
  VARCHAR_ASC
  VARCHAR_DESC
  ENUM_ASC
  ENUM_DESC
  DOMAIN_ASC
  DOMAIN_DESC
  DOMAIN2_ASC
  DOMAIN2_DESC
  COMPOUND_TYPE_ASC
  COMPOUND_TYPE_DESC
}

type UpdatableView {
  x: Int
  name: String
  description: String
  constant: Int
}

type UpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [UpdatableViewsEdge]
  nodes: [UpdatableView!]
}

type UpdatableViewsEdge {
  cursor: Cursor!
  node: UpdatableView!
}

enum UpdatableViewsOrderBy {
  NATURAL
  X_ASC
  X_DESC
  NAME_ASC
  NAME_DESC
  DESCRIPTION_ASC
  DESCRIPTION_DESC
  CONSTANT_ASC
  CONSTANT_DESC
}

scalar Uuid
"
`;

exports[`test schema 2`] = `
"schema {
  query: Query
}

scalar AnInt

scalar AnotherInt

enum Color {
  RED
  GREEN
  BLUE
}

type CompoundKey implements Node {
  id: ID!
  personId2: Int!
  personId1: Int!
  extra: Boolean
  personByPersonId1: Person
  personByPersonId2: Person
}

type CompoundKeysConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [CompoundKeysEdge]
  nodes: [CompoundKey!]
}

type CompoundKeysEdge {
  cursor: Cursor!
  node: CompoundKey!
}

enum CompoundKeysOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  PERSON_ID_2_ASC
  PERSON_ID_2_DESC
  PERSON_ID_1_ASC
  PERSON_ID_1_DESC
  EXTRA_ASC
  EXTRA_DESC
}

type CompoundType {
  a: Int
  b: String
  c: Color
  d: Uuid
}

scalar Cursor

scalar Email

interface Node {
  id: ID!
}

type NonUpdatableView {
  column: Int
}

type NonUpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [NonUpdatableViewsEdge]
  nodes: [NonUpdatableView!]
}

type NonUpdatableViewsEdge {
  cursor: Cursor!
  node: NonUpdatableView!
}

enum NonUpdatableViewsOrderBy {
  NATURAL
  COLUMN_ASC
  COLUMN_DESC
}

type PageInfo {
  hasNextPage: Boolean!
  hasPreviousPage: Boolean!
  startCursor: Cursor
  endCursor: Cursor
}

type PeopleConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PeopleEdge]
  nodes: [Person!]
}

type PeopleEdge {
  cursor: Cursor!
  node: Person!
}

enum PeopleOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  NAME_ASC
  NAME_DESC
  ABOUT_ASC
  ABOUT_DESC
  EMAIL_ASC
  EMAIL_DESC
  CREATED_AT_ASC
  CREATED_AT_DESC
}

type Person implements Node {
  id: ID!
  rowId: Int!
  name: String!
  about: String
  email: Email
  createdAt: String
}

type Post implements Node {
  id: ID!
  rowId: Int!
  headline: String!
  body: String
  authorId: Int
  personByAuthorId: Person
}

type PostsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [PostsEdge]
  nodes: [Post!]
}

type PostsEdge {
  cursor: Cursor!
  node: Post!
}

enum PostsOrderBy {
  PRIMARY_KEY_ASC
  PRIMARY_KEY_DESC
  NATURAL
  ID_ASC
  ID_DESC
  HEADLINE_ASC
  HEADLINE_DESC
  BODY_ASC
  BODY_DESC
  AUTHOR_ID_ASC
  AUTHOR_ID_DESC
}

type Query {
  node(id: ID!): Node
  allNonUpdatableViews(orderBy: NonUpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): NonUpdatableViewsConnection
  allPosts(orderBy: PostsOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PostsConnection
  post(id: ID!): Post
  postByRowId(rowId: Int!): Post
  allTypes(orderBy: TypesOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): TypesConnection
  allUpdatableViews(orderBy: UpdatableViewsOrderBy = NATURAL, before: Cursor, after: Cursor, first: Int, last: Int): UpdatableViewsConnection
  allCompoundKeys(orderBy: CompoundKeysOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): CompoundKeysConnection
  compoundKey(id: ID!): CompoundKey
  compoundKeyByPersonId1AndPersonId2(personId1: Int!, personId2: Int!): CompoundKey
  allPeople(orderBy: PeopleOrderBy = PRIMARY_KEY_ASC, before: Cursor, after: Cursor, first: Int, last: Int): PeopleConnection
  person(id: ID!): Person
  personByEmail(email: Email): Person
  personByRowId(rowId: Int!): Person
}

type Types {
  bigint: Int
  boolean: Boolean
  varchar: String
  enum: Color
  domain: AnInt
  domain2: AnotherInt
  compoundType: CompoundType
}

type TypesConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [TypesEdge]
  nodes: [Types!]
}

type TypesEdge {
  cursor: Cursor!
  node: Types!
}

enum TypesOrderBy {
  NATURAL
  BIGINT_ASC
  BIGINT_DESC
  BOOLEAN_ASC
  BOOLEAN_DESC
  VARCHAR_ASC
  VARCHAR_DESC
  ENUM_ASC
  ENUM_DESC
  DOMAIN_ASC
  DOMAIN_DESC
  DOMAIN2_ASC
  DOMAIN2_DESC
  COMPOUND_TYPE_ASC
  COMPOUND_TYPE_DESC
}

type UpdatableView {
  x: Int
  name: String
  description: String
  constant: Int
}

type UpdatableViewsConnection {
  pageInfo: PageInfo!
  totalCount: Int
  edges: [UpdatableViewsEdge]
  nodes: [UpdatableView!]
}

type UpdatableViewsEdge {
  cursor: Cursor!
  node: UpdatableView!
}

enum UpdatableViewsOrderBy {
  NATURAL
  X_ASC
  X_DESC
  NAME_ASC
  NAME_DESC
  DESCRIPTION_ASC
  DESCRIPTION_DESC
  CONSTANT_ASC
  CONSTANT_DESC
}

scalar Uuid
"
`;
